# Issue 9955: Rational(3)%Rational(-1) fails

archive/issues_009954.json:
```json
{
    "body": "This is inconsistent\n\n```\nsage: Rational(3)%Rational(-1)\nZeroDivisionError: Inverse does not exist.\n```\n\nbut\n\n```\nsage: 3%(-1)\n0\n```\n\n**Assignee:** @aghitza\n\n**Author:** Andr\u00e9 Apitzsch\n\n**Reviewer:** John Cremona\n\n**Merged:** sage-4.6.2.alpha4\n\nIssue created by migration from https://trac.sagemath.org/ticket/9955\n\n",
    "closed_at": "2011-02-07T08:13:53Z",
    "created_at": "2010-09-20T18:20:29Z",
    "labels": [
        "component: basic arithmetic",
        "bug"
    ],
    "milestone": "https://github.com/sagemath/sage/milestones/sage-4.6.2",
    "title": "Rational(3)%Rational(-1) fails",
    "type": "issue",
    "url": "https://github.com/sagemath/sage/issues/9955",
    "user": "https://github.com/haraldschilly"
}
```
This is inconsistent

```
sage: Rational(3)%Rational(-1)
ZeroDivisionError: Inverse does not exist.
```

but

```
sage: 3%(-1)
0
```

**Assignee:** @aghitza

**Author:** André Apitzsch

**Reviewer:** John Cremona

**Merged:** sage-4.6.2.alpha4

Issue created by migration from https://trac.sagemath.org/ticket/9955





---

archive/issue_comments_092095.json:
```json
{
    "body": "<a id='comment:1'></a>\nThis is caused by the following simpler bug:\n\n```\n\nsage: a=Integer(3)\nsage: b=Integer(-1)\nsage: a.inverse_mod(b)\n---------------------------------------------------------------------------\nZeroDivisionError                         Traceback (most recent call last)\n...\n\nZeroDivisionError: Inverse does not exist.\n```\nwhich is easy to fix.  In `sage.rings.integer.Integer.inverse_mod` there is special case for modulus n=1 but not for -1.  Either ass this special case, or replace n by abs(n).",
    "created_at": "2010-12-21T23:23:34Z",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/9955#issuecomment-92095",
    "user": "https://github.com/JohnCremona"
}
```

<a id='comment:1'></a>
This is caused by the following simpler bug:

```

sage: a=Integer(3)
sage: b=Integer(-1)
sage: a.inverse_mod(b)
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
...

ZeroDivisionError: Inverse does not exist.
```
which is easy to fix.  In `sage.rings.integer.Integer.inverse_mod` there is special case for modulus n=1 but not for -1.  Either ass this special case, or replace n by abs(n).



---

archive/issue_events_075168.json:
```json
{
    "actor": "https://github.com/a-andre",
    "created_at": "2011-01-27T11:08:31Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "label": "needs review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/9955#event-75168"
}
```



---

archive/issue_events_075169.json:
```json
{
    "actor": "https://github.com/JohnCremona",
    "created_at": "2011-01-27T17:05:48Z",
    "event": "unlabeled",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "label": "needs review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/9955#event-75169"
}
```



---

archive/issue_events_075170.json:
```json
{
    "actor": "https://github.com/JohnCremona",
    "created_at": "2011-01-27T17:05:48Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "label": "needs work",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/9955#event-75170"
}
```



---

archive/issue_comments_092096.json:
```json
{
    "body": "<a id='comment:3'></a>\nThe patch looks right and I tested that it works (but did not yet test the whole library).\n\nBUT you need to add a doctest to show that the bug has been fixed.  There's a similar doctest in the same function, so just add something similar.\n\nThen I'll look at it again.",
    "created_at": "2011-01-27T17:05:48Z",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/9955#issuecomment-92096",
    "user": "https://github.com/JohnCremona"
}
```

<a id='comment:3'></a>
The patch looks right and I tested that it works (but did not yet test the whole library).

BUT you need to add a doctest to show that the bug has been fixed.  There's a similar doctest in the same function, so just add something similar.

Then I'll look at it again.



---

archive/attachments_013408.json:
```json
{
    "asset_content_type": "application/octet-stream",
    "asset_name": "trac_9955.patch",
    "asset_url": "tarball://root/attachments/ticket9955/trac_9955.patch",
    "created_at": "2011-01-28T08:42:14Z",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "type": "attachment",
    "url": "https://github.com/sagemath/sage/files/ticket9955/trac_9955.patch",
    "user": "https://github.com/a-andre"
}
```



---

archive/issue_comments_092097.json:
```json
{
    "body": "**Attachment:** [trac_9955.patch](https://github.com/sagemath/sage/files/ticket9955/trac_9955.patch)",
    "created_at": "2011-01-28T08:42:14Z",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/9955#issuecomment-92097",
    "user": "https://github.com/a-andre"
}
```

**Attachment:** [trac_9955.patch](https://github.com/sagemath/sage/files/ticket9955/trac_9955.patch)



---

archive/issue_events_075171.json:
```json
{
    "actor": "https://github.com/a-andre",
    "created_at": "2011-01-28T08:44:41Z",
    "event": "unlabeled",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "label": "needs work",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/9955#event-75171"
}
```



---

archive/issue_events_075172.json:
```json
{
    "actor": "https://github.com/a-andre",
    "created_at": "2011-01-28T08:44:41Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "label": "needs review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/9955#event-75172"
}
```



---

archive/issue_comments_092098.json:
```json
{
    "body": "<a id='comment:4'></a>\ndoctest added",
    "created_at": "2011-01-28T08:44:41Z",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/9955#issuecomment-92098",
    "user": "https://github.com/a-andre"
}
```

<a id='comment:4'></a>
doctest added



---

archive/issue_comments_092099.json:
```json
{
    "body": "**Author:** Andr\u00e9 Apitzsch",
    "created_at": "2011-01-28T08:44:41Z",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/9955#issuecomment-92099",
    "user": "https://github.com/a-andre"
}
```

**Author:** André Apitzsch



---

archive/issue_comments_092100.json:
```json
{
    "body": "**Reviewer:** John Cremona",
    "created_at": "2011-01-28T20:45:28Z",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/9955#issuecomment-92100",
    "user": "https://github.com/JohnCremona"
}
```

**Reviewer:** John Cremona



---

archive/issue_comments_092101.json:
```json
{
    "body": "<a id='comment:5'></a>\nReplying to [@a-andre](#comment%3A4):\n> doctest added\n\nThanks!  Positive review.",
    "created_at": "2011-01-28T20:45:28Z",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/9955#issuecomment-92101",
    "user": "https://github.com/JohnCremona"
}
```

<a id='comment:5'></a>
Replying to [@a-andre](#comment%3A4):
> doctest added

Thanks!  Positive review.



---

archive/issue_events_075173.json:
```json
{
    "actor": "https://github.com/JohnCremona",
    "created_at": "2011-01-28T20:45:28Z",
    "event": "unlabeled",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "label": "needs review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/9955#event-75173"
}
```



---

archive/issue_events_075174.json:
```json
{
    "actor": "https://github.com/JohnCremona",
    "created_at": "2011-01-28T20:45:28Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "label": "positive review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/9955#event-75174"
}
```



---

archive/issue_comments_092102.json:
```json
{
    "body": "**Merged:** sage-4.6.2.alpha4",
    "created_at": "2011-02-07T08:13:53Z",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/9955#issuecomment-92102",
    "user": "https://github.com/jdemeyer"
}
```

**Merged:** sage-4.6.2.alpha4



---

archive/issue_events_075175.json:
```json
{
    "actor": "https://github.com/jdemeyer",
    "created_at": "2011-02-07T08:13:53Z",
    "event": "unlabeled",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "label": "positive review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/9955#event-75175"
}
```



---

archive/issue_events_075176.json:
```json
{
    "actor": "https://github.com/jdemeyer",
    "created_at": "2011-02-07T08:13:53Z",
    "event": "closed",
    "issue": "https://github.com/sagemath/sage/issues/9955",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/9955#event-75176"
}
```
